src/lem/mod.rs

```diff
@@ -169,123 +167,213 @@ impl LEMOP {
                 Ok(Self::Hash2(
                     MetaPtr(new_name),
                     *tag,
-                    [MetaPtr(src0_path.clone()), MetaPtr(src1_path.clone())],
+                    [MetaPtr(src0_path), MetaPtr(src1_path)],
                 ))
             }
             Self::Hash3(tgt, tag, src) => {
-                let Some(src0_path) = dmap.get(src[0].name()) else {
+                let Some(src0_path) = map.get(src[0].name()).cloned() else {
                     bail!("{} not defined", src[0].name());
                 };
-                let Some(src1_path) = dmap.get(src[1].name()) else {
+                let Some(src1_path) = map.get(src[1].name()).cloned() else {
                     bail!("{} not defined", src[1].name());
                 };
-                let Some(src2_path) = dmap.get(src[2].name()) else {
+                let Some(src2_path) = map.get(src[2].name()).cloned() else {
                     bail!("{} not defined", src[2].name());
                 };
-                let new_name = format!("{}/{}", path, tgt.name());
-                if dmap.insert(tgt.name().clone(), new_name.clone()).is_some() {
+                let new_name = format!("{}.{}", path, tgt.name());
+                if map.insert(tgt.name().clone(), new_name.clone()).is_some() {
                     bail!("{} already defined", tgt.name());
                 };
                 Ok(Self::Hash3(
                     MetaPtr(new_name),
                     *tag,
-                    [
-                        MetaPtr(src0_path.clone()),
-                        MetaPtr(src1_path.clone()),
-                        MetaPtr(src2_path.clone()),
-                    ],
+                    [MetaPtr(src0_path), MetaPtr(src1_path), MetaPtr(src2_path)],
                 ))
             }
             Self::Hash4(tgt, tag, src) => {
-                let Some(src0_path) = dmap.get(src[0].name()) else {
+                let Some(src0_path) = map.get(src[0].name()).cloned() else {
                     bail!("{} not defined", src[0].name());
                 };
-                let Some(src1_path) = dmap.get(src[1].name()) else {
+                let Some(src1_path) = map.get(src[1].name()).cloned() else {
                     bail!("{} not defined", src[1].name());
                 };
-                let Some(src2_path) = dmap.get(src[2].name()) else {
+                let Some(src2_path) = map.get(src[2].name()).cloned() else {
                     bail!("{} not defined", src[2].name());
                 };
-                let Some(src3_path) = dmap.get(src[3].name()) else {
+                let Some(src3_path) = map.get(src[3].name()).cloned() else {
                     bail!("{} not defined", src[3].name());
                 };
-                let new_name = format!("{}/{}", path, tgt.name());
-                if dmap.insert(tgt.name().clone(), new_name.clone()).is_some() {
+                let new_name = format!("{}.{}", path, tgt.name());
```

Comment:
As discussed recently, the name will likely eventually want to be derived from the canonical names of the source elements. Of course, since you are working with hashes (which know nothing of 'pointers' in their preimages) you won't be able to use that approach directly. I make this note just as a handle into the previous discussion and an observation that this further wrinkle exists.